Cmdline/Action/Replace.cs

```diff
@@ -0,0 +1,142 @@
+                                log.InfoFormat("Replacement {0} {1} found for {2} {3}",
+                                    replacement.ReplaceWith.identifier, replacement.ReplaceWith.version,
+                                    replacement.ToReplace.identifier, replacement.ToReplace.version);
+                                to_replace.Add(replacement);
+                            }
+                        }
+                        catch (ModuleNotFoundKraken)
+                        {
+                            log.InfoFormat("{0} is installed, but it or its replacement is not in the registry",
+                                mod.Key);
+                        }
+                    }
+                }
+            }
+            else
+            {
+                foreach (string mod in options.modules)
+                {
+                    try
+                    {
+                        log.DebugFormat("Checking that {0} is installed", mod);
+                        CkanModule modToReplace = registry.GetInstalledVersion(mod);
+                        if ( modToReplace != null) 
+                        {
+                            log.DebugFormat("Testing {0} {1} for possible replacement", modToReplace.identifier, modToReplace.version);
+                            try
+                            {
+                                // Check if replacement is available
+                                ModuleReplacement replacement = registry.GetReplacement(mod, ksp.VersionCriteria());
+                                if (replacement != null)
+                                {
+                                    // Replaceable
+                                    log.InfoFormat("Replacement {0} {1} found for {2} {3}",
+                                        replacement.ReplaceWith.identifier, replacement.ReplaceWith.version,
+                                        replacement.ToReplace.identifier, replacement.ToReplace.version);
+                                    to_replace.Add(replacement);
+                                }
+                                log.InfoFormat("Attempt to replace {0} failed, replacement {1} is not compatible",
+                                    mod, modToReplace.replaced_by.name);
+                            }
+                            catch (ModuleNotFoundKraken)
+                            {
+                                log.InfoFormat("{0} is installed, but its replacement {1} is not in the registry",
+                                    mod, modToReplace.replaced_by.name);
+                            }
+                        }
+                    }
+                    catch (ModuleNotFoundKraken kraken)
+                    {
+                        User.RaiseMessage("Module {0} not found", kraken.module);
+                    }                                           
+                }
+            }
+            // TODO: These instances all need to go.
```

Comment:
Any idea what this comment means? I see it in `Upgrade.cs`, but if we don't know what change it is suggesting, we might as well not carry it forward.

Reply:
It is about the `GetInstance(ksp, User)` stuff below. It's ugly as hell, but it's pretty endemic throughout the code due to someone's possibly unnecessary decisions in the data design. I think the original idea was to have different KSP instances/registries all loaded in memory  at the same time, and allow the user to switch quickly between them. At some stage, that became infeasible, but the fundamental structure will take a significant project to remove.